src/include/libpmemobj.h

```diff
@@ -84,6 +99,553 @@ void pmemobj_set_funcs(
+ */
+PMEMoid pmemobj_get_root(PMEMobjpool *pop, size_t size);
+
+/*
+ * Returns the size in bytes of the root object. Always equal to the requested
+ * size.
+ */
+size_t pmemobj_sizeof_root(PMEMobjpool *pop);
+
+/*
+ * The following set of macros and functions allow access to the entire
+ * collection of objects, or objects of given type.
+ *
+ * Use with conjunction with non-transactional allocations. Pmemobj pool acts
+ * as a generic container (list) of objects that are not assigned to any
+ * user-defined data structures.
+ */
+
+/*
+ * Returns the first object of the specified type number.
+ */
+PMEMoid pmemobj_get_first(PMEMobjpool *pop, int type_num);
+
+/*
+ * Returns the next object (in order of allocations) of the same type.
+ */
+PMEMoid pmemobj_get_next(PMEMoid oid);
+
+/*
+ * Iterates through every existing allocated object.
+ */
+#define	PMEMOBJ_FOREACH(oidvar, vartype_num, pop)			\
+for (vartype_num = 0; vartype_num < MAX_OID_TYPE_NUM; ++vartype_num)	\
+	for (oidvar = pmemobj_get_first(pop, vartype_num);		\
+		(oidvar).off != 0; oidvar = pmemobj_get_next(oidvar))
+
+/*
+ * Iterates through every object of the specified type number.
+ */
+#define	PMEMOBJ_FOREACH_TYPE(var, pop, type_num)			\
+for (OID_ASSIGN(var, pmemobj_get_first(pop, type_num));			\
+		OID_IS_NULL(var) == 0;					\
+		OID_ASSIGN(var, pmemobj_get_next((var).oid)))
+
+/*
+ * Non-transactional persistent atomic circular doubly-linked list
+ */
+#define	PLIST_ENTRY(type)						\
+struct {								\
+	OID_TYPE(type) pe_next;						\
+	OID_TYPE(type) pe_prev;						\
+	PMEMmutex lock;							\
+}
+
+#define	PLIST_HEAD(name, type)						\
+struct name {								\
+	OID_TYPE(type) pe_first;					\
+	PMEMmutex lock;							\
+}
+
+int pmemobj_list_insert(PMEMobjpool *pop, size_t pe_offset, PMEMoid dest,
+	PMEMoid *head, int before, PMEMoid oid);
+
+int pmemobj_list_insert_new(PMEMobjpool *pop, size_t pe_offset, PMEMoid dest,
+	PMEMoid *head, int before, size_t size, int type_num);
+
+int pmemobj_list_remove(PMEMobjpool *pop, size_t pe_offset, PMEMoid oid,
+	PMEMoid *head, int free);
+
+int pmemobj_list_move(PMEMobjpool *pop,
+	size_t pe_old_offset, size_t pe_new_offset,
+	PMEMoid *head_old, PMEMoid *head_new, int tail, PMEMoid oid);
+
+#define	PLIST_FIRST(head)	((head)->pe_first)
+#define	PLIST_LAST(head, field)	(D_RO((head)->pe_first)->field.pe_prev)
+#define	PLIST_EMPTY(head)	(OID_IS_NULL((head)->pe_first))
+#define	PLIST_NEXT(elm, field)	(D_RO(elm)->field.pe_next)
+#define	PLIST_PREV(elm, field)	(D_RO(elm)->field.pe_prev)
+
+#define	PLIST_FOREACH(var, head, field)					\
+for (TYPED_OID_ASSIGN((var), PLIST_FIRST((head)));			\
+	OID_IS_NULL((var)) == 0;					\
+	TYPED_OID_ASSIGN((var), PLIST_NEXT((var), field)))
+
+#define	PLIST_FOREACH_REVERSE(var, head, field)				\
+for (TYPED_OID_ASSIGN((var), PLIST_LAST((head), field));		\
+	OID_IS_NULL((var)) == 0;					\
+	TYPED_OID_ASSIGN((var), PLIST_PREV((var), field)))
+
+#define	PLIST_INSERT_HEAD(pop, head, elm, field)			\
```

Comment:
Same comment as for *insert* functions. Do we need all those variants? Why not just one *PLIST_INSERT()* macro?